File: python/repl.py
from __future__ import annotations
from errors import LangError
from langtypes import LangString, fmtStackPrint
"""
	repl - run code interactively, run unittests, or run programs.

	Copyright (c) 2022 Frank McIngvale, see LICENSE

	Ported from C++ implementation
"""
from interpreter import Interpreter
import sys, re, os

INITLIB = "../lib/init.verb.b"
COMPILERLIB = "../lib/compiler.verb.b"
PATCHESLIB = "../lib/patches.verb"

def compile_and_load(intr, text, allow_overwrite):
	import native
	native.ALLOW_OVERWRITE_WORDS = allow_overwrite

	# normally do NOT want to catch errors here since it's better for them to be
	# caught in the normal flow. however sometimes a compiler bug happens where this
	# needs to be uncommented see what it did

	#try:
	# push code, compile and load into interpreter
	intr.push(LangString(text))
	code = intr.lookupWordOrFail('compile-and-load-string')
	intr.run(code)
	#except LangError as exc:
	#	errmsg = "*** " + exc.msg + " ***"
	#	print(errmsg)
	#	# always backtrace here since this is an error in the compiler not user code
	#	print_backtrace(intr)
	#	sys.exit(1) # treat as fatal
		
	native.ALLOW_OVERWRITE_WORDS = False # set back to default

def deserialize_and_run(intr, filename):
	from deserialize import deserialize_stream
	fileIn = open(filename, "r")
	deserialize_stream(intr, fileIn)
	code = intr.lookupWordOrFail("__main__")
	intr.run(code)
	# always delete __main__ after running, else it will prevent other code from loading
	intr.deleteWord("__main__")
	
def debug_hook(intr: Interpreter, word: str):
	print("=> " + intr.reprStack())
	print("Run: " + str(word))
	sys.stdout.write("press ENTER to continue ...")
	sys.stdout.flush()
	sys.stdin.readline()

def compile_and_run(intr, text, singlestep, allow_overwrite=False):
	# push code, compile and load into interpreter
	compile_and_load(intr, text, allow_overwrite)
	
	# run __main__
	code = intr.lookupWordOrFail('__main__')

	if singlestep:
		intr.run(code,debug_hook)
	else:
		intr.run(code)

	# as above, delete __main__ to avoid inadvertent reuse
	intr.deleteWord("__main__")

def new_interpreter(verbose=False):
	"convenience to start interpreter and optionally load init lib"
	intr = Interpreter()

	# load serialized versions of init.verb and compiler.verb (required
	# to bootstrap interpreter)
	deserialize_and_run(intr, INITLIB)
	deserialize_and_run(intr, COMPILERLIB)
	
	# now load & run patches, allowing them to overwrite existing words
	# (this is the only place this is allowed)
	buf = open(PATCHESLIB,'r').read()
	# if this is likely to take a bit, print a message
	if verbose and len(buf) > 1000: print("Patching ...")
	
	compile_and_run(intr, buf, False, True)

	return intr

# returns string on error, None on success	
def safe_compile_and_run(intr, text, singlestep, backtrace_on_error):
	try:
		compile_and_run(intr, text, singlestep)
	except LangError as exc:
		errmsg = "*** " + exc.msg + " ***"
		if backtrace_on_error:
			print_backtrace(intr)
		
		return errmsg

def repl(singlestep, showstats):
	"Run interactively"
	print("Verbii running on Python {0}.{1}.{2}".format(sys.version_info.major, sys.version_info.minor, sys.version_info.micro))

	intr = new_interpreter(True)

	while(True):
		sys.stdout.write(">> ")
		sys.stdout.flush()
		line = sys.stdin.readline().strip()
		if len(line) == 0:
			return # eof
		if line == "quit" or line == ",q":
			if showstats: intr.print_stats()
			return
		
		err = safe_compile_and_run(intr, line, singlestep, True)
		if err is not None:
			print(err)
			intr = new_interpreter() # restart on error
		else:
			print("=> " + intr.reprStack())

def run_test_mode(filename: str):
	"""read one line at a time from file and run, printing results and stack. 
	used for unit testing"""
	intr = new_interpreter()

	re_blankline = re.compile(r"""(^[ \t\r\n]*$)""")
	fileIn = open(filename,'r')
	while True:
		line = fileIn.readline()
		if line == "": return # end of file
		if re_blankline.match(line):
			continue # skip blank lines
		
		line = line.rstrip()
		print(">> " + line) 

		# don't want backtraces in -test mode; if an unknown error occurs, rerun
		# case without -test to see backtrace
		err = safe_compile_and_run(intr, line, False, False)
		if err is not None:
			print(err)
			intr = new_interpreter() # restart on error
		else:	
			print("=> " + intr.reprStack())
		
def backtrace_curframe(intr: Interpreter):
	trace = ""
	nr = 7; # number of words to print in each frame
	while nr > 0:
		w = intr.prevCodeObject()
		if(w == None):
			print(trace)
			return
		else:
			trace = fmtStackPrint(w) + ' ' + trace
		
		nr -= 1
	
	print(trace)

def print_backtrace(intr: Interpreter):
	i=0
	while True:
		sys.stdout.write("FRAME " + str(i) + ": ")
		i += 1
		backtrace_curframe(intr)
		if intr.havePushedFrames():
			intr.code_return() # pop frame and print next
		else:
			return # end of callstack

def run_file(intr: Interpreter, filename: str, singlestep: bool, showstats: bool):
	# run file
	buf = open(filename,'r').read()

	err = safe_compile_and_run(intr, buf, singlestep, True)
	if err is not None:
		print(err)
	elif showstats:
		intr.print_stats()

if __name__ == '__main__':
	testmode = False
	filename = None
	singlestep = False
	showstats = False
	for i,arg in enumerate(sys.argv[1:]):
		#print("ARG:",arg)
		if arg == '-test':
			testmode = True
		elif arg == '-step':
			singlestep = True
		elif arg == '-stats':
			showstats = True
		elif arg == '--':
			# pass rest of args to script
			import native
			# this will be pushed directly, so make into list of string objects
			native.NATIVE_CMDLINE_ARGS = []
			for narg in sys.argv[i+2:]:
				native.NATIVE_CMDLINE_ARGS.append(LangString(narg))

			#print("CMDLINE ARGS TO SCRIPT:",sys.argv[i+2:])
			break
		elif filename is None and os.path.exists(arg):
			filename = arg
		else:
			print("Unknown option: " + arg)
			sys.exit(1)

	if filename is None:
		repl(singlestep, showstats)
	elif testmode is True:
		run_test_mode(filename)
	else:
		intr = new_interpreter()
		run_file(intr, filename, singlestep, showstats)
		

File: lua/repl.lua
--[[
	repl - run code interactively, run unittests, or run programs.

	Copyright (c) 2022 Frank McIngvale, see LICENSE

	Ported from Python implementation
]]

require("deserialize")
require("interpreter")
require("langtypes")
require("native")

local INITLIB = "../lib/init.verb.b"
local COMPILERLIB = "../lib/compiler.verb.b"
local PATCHESLIB = "../lib/patches.verb"

local SHOW_RUNTIME_STATS = false

function readfile(filename)
	local f = io.open(filename, "r")
	if f == nil then
		error(">>>No such file: " .. filename)
	end
	local buf = f:read("a")
	io.close(f)
	return buf
end

function compile_and_load(intr, text, allow_overwrite)
	ALLOW_OVERWRITING_WORDS = allow_overwrite
	intr:push(new_String(text))
	code = intr:lookupWordOrFail('compile-and-load-string')
	intr:run(code)
	-- turn flag back off (default)
	ALLOW_OVERWRITING_WORDS = false
end

function deserialize_and_run(intr, filename)
	local f = io.open(filename, "r")
	deserialize_stream(intr, f)
	io.close(f)
	local code = intr:lookupWordOrFail('__main__')
	intr:run(code)
	-- always delete __main__ after running, or next file will fail to load
	intr:deleteWord('__main__')
end

function debug_hook(intr, word)
	print("=> " .. intr:reprStack())
	print("Run: " .. word)
	io.write("press ENTER to continue ...")
	io.flush()
	io.read()
end

-- use safe_ version instead - this has no error checking
function compile_and_run(intr, text, singlestep, allow_overwrite)
	if allow_overwrite == null then allow_overwrite = false end

	compile_and_load(intr, text, allow_overwrite)
	
	-- now run the __main__ that was compiled
	local code = intr:lookupWordOrFail('__main__')
	
	if singlestep then
		intr:run(code,debug_hook)
	else
		intr:run(code)
	end
	
	-- remove __main__ when done
	intr:deleteWord('__main__')
end

function make_interpreter(verbose)
	local intr = new_Interpreter()
	
	-- load bootstrap libraries so compiler works
	deserialize_and_run(intr, INITLIB)
	deserialize_and_run(intr, COMPILERLIB)
	
	-- load & run patches -- allow patches to overwrite existing words
	local buf = readfile(PATCHESLIB)
	-- if this will take a bit, print a message
	if verbose and #buf > 1000 then print("Patching ...") end
	compile_and_run(intr, readfile(PATCHESLIB), false, true)

	return intr
end

-- returns error string or nil if no error
function safe_compile_and_run(intr, text, singlestep, backtrace_on_error)
	local result,error = pcall(compile_and_run, intr, text, singlestep, false)
	if result == false then
		-- match sequence >>> to strip out filename from error message that lua added
		local match = string.match(error, "^.+>>>")
		if match == nil then
			-- didn't get expected error format, so print raw error to show something at least
			local errstr = "*** " .. error .. " ***"
			return errstr
		else
			local errstr = "*** " .. string.sub(error, #match+1) .. " ***"
			if backtrace_on_error then
				print_backtrace(intr)
			end
			return errstr
		end
	else
		return nil -- no error
	end
end

function repl(singlestep)
	-- Run interactively
	print("Verbii running on " .. _VERSION)
	local intr = make_interpreter(true)

	while true do
		io.write(">> ")
		io.flush()
		local line = io.read()
		if line == nil then
			return -- eof
		elseif line == "quit" or line == ",q" then
			if SHOW_RUNTIME_STATS then
				intr:print_stats()
			end
			return
		end
		
		-- compile & run each line, watching for errors
		local err = safe_compile_and_run(intr, line, singlestep, true)
		if err ~= nil then
			print(err)
			intr = make_interpreter() -- restart interpreter on error
		else
			print("=> " .. intr:reprStack())
		end
	end
end

function run_test_mode(filename)
	-- read one line at a time from file and run, printing results and stack. 
	-- used for unit testing
	local intr = make_interpreter(noinit)

	local fileIn = io.open(filename,"r")
	while true do
		::LOOP::
		local line = fileIn:read("l") -- discard \n at end of line
		if not line then
			break
		end
		if string.match(line, "^[%s]*$") then
			goto LOOP -- skip blank lines
		end

		print(">> " .. line)
		
		-- i only want the errors not the backtraces -- if an unexpected error occurred,
		-- just run again in non-test mode to see backtrace
		local err = safe_compile_and_run(intr, line, false, false)
		if err ~= nil then
			print(err)
			intr = make_interpreter() -- restart interpreter on error
		else
			print("=> " .. intr:reprStack())
		end
	end
	io.close(fileIn)
end

function backtrace_curframe(intr)
	local trace = ""
	local nr = 7 -- number of words to print in each frame
	while nr > 0 do
		local w = intr:prevObj()
		if w == nil then
			print(trace)
			return
		else
			trace = fmtStackPrint(w) .. " " .. trace
		end
		nr = nr - 1
	end
	
	print(trace)
end

function print_backtrace(intr)
	local i=0
	while true do
		io.write("FRAME " .. tostring(i) .. ": ")
		i = i + 1
		backtrace_curframe(intr)
		if intr:havePrevFrames() then
			intr:code_return() -- pop current frame and print next one
		else
			return -- end of callstack, done
		end
	end
end

function run_file(intr, filename, singlestep)
	-- load & run file
	local buf = readfile(filename)
	local err = safe_compile_and_run(intr, buf, singlestep, true)
	if err ~= nil then
		print(err)
	elseif SHOW_RUNTIME_STATS then
		intr:print_stats()
	end
end

local filename = nil
local noinit = false
local test_mode = false
local singlestep = false
local args_to_script = {}

for i=1,#arg do
	if arg[i] == "-noinit" then
		noinit = true
	elseif arg[i] == "-test" then
		test_mode = true
	elseif arg[i] == "-step" then
		singlestep = true
	elseif arg[i] == "-stats" then
		SHOW_RUNTIME_STATS = true
	elseif arg[i] == "--" then
		-- rest of args go to script
		i = i + 1
		while i <= #arg do
			table.insert(args_to_script, new_String(arg[i]))
			i = i + 1
		end
		break
	elseif filename == nil then
		filename = arg[i]
	else
		error(">>>Bad command line argument: " .. arg[i])
	end
end

set_native_cmdline_args(args_to_script)

if filename == nil then
	repl(singlestep)
elseif test_mode then
	run_test_mode(filename)
else
	local intr = make_interpreter(noinit)
	run_file(intr, filename, singlestep)
end
